WooCommerce Products Sync: Add product title, status, type, slug, and dates

chihsuan opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes WOOA7S-321


## Proposed changes:
<!--- Explain what functional changes your PR includes -->

Enhanced the WooCommerce Products Sync module to include post-related fields alongside existing product lookup data.

- Added `title`, `post_status`, `type`, `slug`, `date_created`, and `date_modified` fields to sync payload

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

This PR expands the WooCommerce product data that gets synced to include post-related fields (title, status, type, slug, creation/modification dates). The additional data helps provide more complete product information for synchronization purposes.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->


* Set up a Site with WooCommerce and this JP branch
* Enable the sync module by adding the filter `add_filter( 'jetpack_sync_modules', array( __CLASS__, 'add_woocommerce_products_sync_module' ) );` in `projects/packages/sync/src/class-actions.php`
* Follow the test instructions in 190832-ghe-Automattic/wpcom

chihsuan committed:
Enhance WooCommerce Products Sync: Add post-related fields and datetime conversion method. Update transform_product_data to include product title, status, type, and creation/modification dates. Implement datetime_to_object for proper encoding of WC_DateTime objects.

chihsuan committed:
Add changelog

Nikschavan commented on projects/packages/sync/src/modules/class-woocommerce-products.php:
```diff
@@ -353,4 +361,26 @@ public function get_next_chunk( $config, $status, $chunk_size ) {
+			}
+			return (object) (array) $wc_datetime;
+		}
+	}
```
```suggestion
		}
		
		return null;
	}
```

Nitpick, feel free to ignore: Adding an explicit return statement.

Nikschavan commented on projects/packages/sync/src/modules/class-woocommerce-products.php:
```diff
@@ -268,26 +270,32 @@ public function get_product_by_ids( $ids, $order = '' ) {
-		$product_id = $product_data['product_id'];
-
-		// Attempt to retrieve the WooCommerce product object and its COGS value.
-		$product_data['cogs_amount'] = null;
```
The default value for `cogs_amount` is no longer set. Is this expected, or do we explicitly want to set these array keys to `null`? I see that on the server side, these are automatically set to their default value types anyway, so I _think_  will not cause any PHP warnings

chihsuan replied:
Yes, it's okay to leave them unset.

Nikschavan commented on projects/packages/sync/src/modules/class-woocommerce-products.php:
```diff
@@ -268,26 +270,32 @@ public function get_product_by_ids( $ids, $order = '' ) {
+			$product_data['type']          = $product->get_type();
+			$product_data['slug']          = $product->get_slug();
+			$product_data['date_created']  = $this->datetime_to_object( $product->get_date_created() );
+			$product_data['date_modified'] = $this->datetime_to_object( $product->get_date_modified() );
```
While rare, `WC_DateTime` could throw an exception if the date or timezone is incorrect (Both values are returned from the database) - Should we try/catch the call to datetime_to_object()?

chihsuan replied:
Thanks for bringing this up! I checked the WooCommerce Core codebase, and it doesn't use try/catch blocks in similar code. This code is also used in WooCommerce Analytics, so I think it’s safe to assume that the dates are valid.

Nikschavan reviewed (dismissed):
Looks great, added some non-blocking suggestions

darssen reviewed (commented):
Although by briefly looking at the changes, I don't see anything explicitly wrong, I see that we are now instantiating the product every time. 
This was raised as a concern [here](https://github.com/Automattic/jetpack/pull/44601#pullrequestreview-3099847307) when it was more limited for cogs, but now it seems we want to do it every time. How might it impact performance for big sites or sites that change products a lot?

chihsuan committed:
Update projects/packages/sync/src/modules/class-woocommerce-products.php

Co-authored-by: Nikhil <Nikschavan@users.noreply.github.com>

chihsuan committed:
Refactor WooCommerce Products Sync

Streamline product data retrieval by introducing methods for fetching product posts, meta data, and COGS values. Enhance get_product_by_ids to consolidate product information and improve performance. Remove deprecated code and ensure compatibility with WooCommerce product types.

chihsuan committed:
Rename products_data -> products

Nikschavan commented on projects/packages/sync/src/modules/class-woocommerce-products.php:
```diff
@@ -353,4 +360,171 @@ public function get_next_chunk( $config, $status, $chunk_size ) {
+				'order'       => $order,
+				'post_type'   => array( 'product', 'product_variation' ),
+				'post_status' => 'any',
+				'numberposts' => -1, // Get all posts.
```
[There could be very large stores](https://pressable.com/blog/scaling-woocommerce-for-large-stores-handling-100000-products/) with 100k+ products. I think loading all posts at once could cause performance issues in such cases. Can we paginate with a safe max number of posts per page, like 100 or something?

Nikschavan replied:
Never mind, it looks like this is already chunked at 100 by Jetpack sync - https://github.com/Automattic/jetpack/blob/c1addf1d5b6bb57e55e874cdaf038cac22196093/projects/packages/sync/src/modules/class-module.php#L427-L431

chihsuan committed:
Clarify purpose of WooCommerce Products sync module

Added notes indicating that the WooCommerce Products sync module is currently used for analytics purposes only in both the Actions class and the WooCommerce_Products module documentation.

chihsuan committed:
Update comment

darssen reviewed (approved):
From my review and testing, things are working fine.

The status of where the module lives now is a bit clumsy, since it will not work without the Analytics plugin. I appreciate the comments in [dbb3b4f](https://github.com/Automattic/jetpack/pull/44810/commits/dbb3b4ffe7eac6c8cb7adc7a0393db3a62069d7b) to explicitly state it. But since it looks like there will be some rework around it, I don't want to block the work here.

I'm also reviewing the analogous wpcom PR, so after that one is done we can merge this one.

chihsuan committed:
Merge branch 'trunk' into update/sync-product-data
